api/src/main/java/org/pmiops/workbench/workspaces/resources/WorkspaceResourceMapper.java

```diff
@@ -128,6 +179,96 @@ default WorkspaceResource fromDbUserRecentResource(
+      DbUserRecentlyModifiedResource dbUserRecentResource,
+      FirecloudWorkspaceResponse fcWorkspace,
+      DbWorkspace dbWorkspace,
+      DbConceptSet dbConceptSet) {
+    return mergeWorkspaceAndResourceFields(
+        fromWorkspace(dbWorkspace),
+        fcWorkspace,
+        fromDbUserRecentlyModifiedResource(dbUserRecentResource, dbConceptSet));
+  }
+
+  default WorkspaceResource fromDbUserRecentlyModifiedResourceAndDataSet(
+      DbUserRecentlyModifiedResource dbUserRecentResource,
+      FirecloudWorkspaceResponse fcWorkspace,
+      DbWorkspace dbWorkspace,
+      DbDataset dbDataset) {
+    return mergeWorkspaceAndResourceFields(
+        fromWorkspace(dbWorkspace),
+        fcWorkspace,
+        fromDbUserRecentlyModifiedResource(dbUserRecentResource, dbDataset));
+  }
+
+  default WorkspaceResource fromDbUserRecentlyModifiedResourceAndNotebookName(
+      DbUserRecentlyModifiedResource dbUserRecentResource,
+      FirecloudWorkspaceResponse fcWorkspace,
+      DbWorkspace dbWorkspace,
+      String notebookName) {
+    return mergeWorkspaceAndResourceFields(
+        fromWorkspace(dbWorkspace),
+        fcWorkspace,
+        fromDbUserRecentlyModifiedResource(dbUserRecentResource, notebookName));
+  }
+
+  default WorkspaceResource fromDbUserRecentlyModifiedResource(
+      DbUserRecentlyModifiedResource dbUserRecentlyModifiedResource,
+      FirecloudWorkspaceResponse fcWorkspace,
+      DbWorkspace dbWorkspace,
+      CohortService cohortService,
+      ConceptSetService conceptSetService,
+      DataSetService dataSetService) {
+    final long workspaceId = dbUserRecentlyModifiedResource.getWorkspaceId();
+    switch (dbUserRecentlyModifiedResource.getResourceType()) {
+      case COHORT:
+        return fromDbUserRecentlyModifiedResourceAndCohort(
+            dbUserRecentlyModifiedResource,
+            fcWorkspace,
+            dbWorkspace,
+            cohortService.findDbCohortByCohortId(
+                getResourceIdInLong(dbUserRecentlyModifiedResource)));
+      case CONCEPT_SET:
+        return fromDbUserRecentlyModifiedResourceAndConceptSet(
+            dbUserRecentlyModifiedResource,
+            fcWorkspace,
+            dbWorkspace,
+            conceptSetService.getDbConceptSet(
+                workspaceId, getResourceIdInLong(dbUserRecentlyModifiedResource)));
+      case DATA_SET:
+        return fromDbUserRecentlyModifiedResourceAndDataSet(
+            dbUserRecentlyModifiedResource,
+            fcWorkspace,
+            dbWorkspace,
+            dataSetService
+                .getDbDataSet(workspaceId, getResourceIdInLong(dbUserRecentlyModifiedResource))
+                .orElse(new DbDataset()));
+      case NOTEBOOK:
+        return fromDbUserRecentlyModifiedResourceAndNotebookName(
+            dbUserRecentlyModifiedResource,
+            fcWorkspace,
+            dbWorkspace,
+            dbUserRecentlyModifiedResource.getResourceId());
+    }
+    return null;
```

Comment:
A null would probably blow up here in a confusing way.

Since it would be an error to reach this code, let's change it to a `default` in the switch, which throws.